python/core/qgsproperty.sip

```diff
@@ -0,0 +1,870 @@
+%End
+
+  public:
+
+    /** Constructor for QgsFieldBasedProperty.
+     * @param field field name
+     * @param isActive whether the property is intially active
+     */
+    QgsFieldBasedProperty( const QString& field = QString(), bool isActive = false );
+
+    virtual Type propertyType() const;
+
+    virtual QgsFieldBasedProperty* clone() /Factory/;
+
+    /** Sets the field name the property references.
+     * @param field field name
+     * @see field()
+     */
+    void setField( const QString& field );
+
+    /** Returns the current field name the property references.
+     * @see setField()
+     */
+    QString field() const;
+
+    virtual QSet< QString > referencedFields( const QgsExpressionContext& context = QgsExpressionContext() ) const;
+
+    bool writeXML( QDomElement& propertyElem, QDomDocument& doc ) const;
+    bool readXML( const QDomElement& propertyElem, const QDomDocument& doc );
+
+  protected:
+
+    virtual QVariant propertyValue( const QgsExpressionContext& context, const QVariant& defaultValue = QVariant() ) const;
+
+};
+
+/** \ingroup core
+ * \class QgsExpressionBasedProperty
+ * \brief QgsAbstractProperty subclass which takes its value from a QgsExpression result.
+ * \note Added in version 2.16
+ */
+
+class QgsExpressionBasedProperty : QgsAbstractProperty
+{
+%TypeHeaderCode
+#include <qgsproperty.h>
+%End
+
+  public:
+
+    /** Constructor for QgsExpressionBasedProperty.
+     * @param expression expression string
+     * @param isActive whether the property is intially active
+     */
+    QgsExpressionBasedProperty( const QString& expression = QString(), bool isActive = false );
+
+    virtual Type propertyType() const;
+
+    virtual QgsExpressionBasedProperty* clone() /Factory/;
+
+    /** Sets the expression to use for the property value.
+     * @param expression expression string
+     * @see expressionString()
+     */
+    void setExpressionString( const QString& expression );
+
+    /** Returns the expression to use for the property value.
+     * @see setExpressionString()
+     */
+    QString expressionString() const;
+
+    virtual QSet< QString > referencedFields( const QgsExpressionContext& context = QgsExpressionContext() ) const;
+
+    bool writeXML( QDomElement& propertyElem, QDomDocument& doc ) const;
+    bool readXML( const QDomElement& propertyElem, const QDomDocument& doc );
+
+  protected:
+
+    virtual QVariant propertyValue( const QgsExpressionContext& context, const QVariant& defaultValue = QVariant() ) const;
+
+};
+
+
+/** \ingroup core
+ * \class QgsPropertyTransformer
+ * \brief Abstract base class for objects which transform the calculated value of a property.
+ * Possible uses include transformers which map a value into a scaled size or color from a gradient.
+ * \note Added in version 2.16
+ */
+
+class QgsPropertyTransformer
+{
+%TypeHeaderCode
+#include <qgsproperty.h>
+%End
+
+%ConvertToSubClassCode
+  if (dynamic_cast<QgsSizeScaleTransformer*>(sipCpp) != NULL)
```

Comment:
And use `nullptr` and spacing like in the C++ source even if it's not enforced by astyle ;)